### PR TITLE
Moved logits `.float()` to loss and compiled it if compiling

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #567
* __->__ #551

Compiling the loss improves performance. Moving the `.float()` upcast to inside this compiled loss further improves performance.

### CODE REVIEW

**Review by yifuwang — APPROVED** [review]

**Comment by awgu** [comment]
@yifuwang I need to fix PP before this is landable 😢

File: train.py [hunk 1]
```diff
@@ -137,7 +137,7 @@ def main(job_config: JobConfig):
     # loss function to be shared by Pipeline Parallel and SPMD training
     def loss_fn(pred, labels):
         return torch.nn.functional.cross_entropy(
-            pred.flatten(0, 1), labels.flatten(0, 1)
+            pred.flatten(0, 1).float(), labels.flatten(0, 1)
```
Comment (awgu) [inline train.py:140]: cc: @Chillee `torch.compile` should respect this upcast numerically?

File: train.py [hunk 2]
```diff
@@ -289,7 +289,10 @@ def loss_fn(pred, labels):
                 # Non-PP forward / backward
                 with train_context():
                     pred = model(input_ids)
-                    loss = loss_fn(pred, labels)
+                    if job_config.training.compile:
+                        loss = torch.compile(loss_fn)(pred, labels)
```
Comment (tianyu-l) [inline train.py:293]: I wonder if this would break Loss Parallel when TP is applied. It shouldn't as `torch.compile` will trace into `__torch_dispatch__` (how Loss Parallel is implemented) but it'd be good to check the memory and numerics.

File: train.py (hunk 1 above)
Comment (Chillee) [reply train.py:140]: yeah definitely

**Comment by wconstab** [comment]
> @H-Huang @wconstab do you have any idea if the output logits being fp32 is a hard requirement for PP? anyway we can leave them as bf16?

sorry didn't see this before.

This is just the outputs of the last layer?  PP should be OK with this, with a few caveats-
1- if its not the last layer, then we'll send/recv it, and iirc we do not have nccl support for 8-bit types, though 16 should be fine.
2- we also need to know the size when we allocate send/recv buffers.  In august this would have had to be configured at PipelineStage.__init__ time when you pass the manual 'input/output'_args values, but recently landed shape-inference runs at runtime based on real tensors provided to .step() api so this should be totally transparent now

Both of those points should not apply to the values produced as outputs of the last stage (or the losses), so i'll need more details about what kind of error you see.
